File: ch8/py/c-interrupt.py
import asyncio
from contextlib import aclosing

from langchain.schema import HumanMessage
from langgraph.graph import StateGraph
from langgraph.checkpoint.memory import MemorySaver


async def main():
    # Create a simple graph
    builder = StateGraph()
    # Add nodes and edges as needed
    graph = builder.compile(checkpointer=MemorySaver())

    event = asyncio.Event()

    input = {
        "messages": [
            HumanMessage(
                "How old was the 30th president of the United States when he died?"
            )
        ]
    }

    config = {"configurable": {"thread_id": "1"}}

    async with aclosing(graph.astream(input, config)) as stream:
        async for chunk in stream:
            if event.is_set():
                break
            else:
                print(chunk)  # do something with the output

    # Simulate interruption after 2 seconds
    await asyncio.sleep(2)
    event.set()


if __name__ == "__main__":
    asyncio.run(main())

File: ch8/js/c-interrupt.ts
import { HumanMessage } from "@langchain/core/messages";
import { Annotation } from "@langchain/langgraph";
import { MemorySaver } from "@langchain/langgraph";
import { StateGraph } from "@langchain/langgraph";
import * as dotenv from "dotenv/config";
dotenv;

const GraphState = Annotation.Root({
  messages: Annotation(),
});

const controller = new AbortController();

const input = {
  messages: [
    new HumanMessage(
      "How old was the 30th president of the United States when he died?"
    ),
  ],
};

const config = { configurable: { thread_id: "1" } };

// Assuming graph is already created and configured
const graph = new StateGraph(GraphState).compile({
  checkpointer: new MemorySaver(),
});

// Simulate interruption after 2 seconds
setTimeout(() => {
  controller.abort();
}, 2000);

try {
  const output = await graph.stream(input, {
    ...config,
    signal: controller.signal,
  });

  for await (const chunk of output) {
    console.log(chunk); // do something with the output
  }
} catch (e) {
  console.log(e);
}
